File: platforms/android/stato/src/main/java/com/facebook/stato/plugins/litho/DataUtils.java
package com.facebook.stato.plugins.litho;

import static com.facebook.stato.plugins.inspector.InspectorValue.Type.Color;

import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import com.facebook.stato.core.StatoObject;
import com.facebook.stato.plugins.inspector.InspectorValue;
import com.facebook.stato.plugins.inspector.Named;
import com.facebook.litho.StateContainer;
import com.facebook.litho.annotations.Prop;
import com.facebook.litho.annotations.State;
import java.lang.reflect.Field;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;

public class DataUtils {

  @Nullable
  static List<com.facebook.stato.plugins.inspector.Named<com.facebook.stato.core.StatoObject>>
      getPropData(Object node) throws Exception {
    final StatoObject.Builder props = new StatoObject.Builder();
    List<Named<StatoObject>> data = new ArrayList<>();

    boolean hasProps = false;

    for (Field f : node.getClass().getDeclaredFields()) {
      f.setAccessible(true);

      final Prop annotation = f.getAnnotation(Prop.class);
      if (annotation != null) {
        if (f.get(node) != null
            && PropWithInspectorSection.class.isAssignableFrom(f.get(node).getClass())) {
          final AbstractMap.SimpleEntry<String, String> datum =
              ((PropWithInspectorSection) f.get(node)).getStatoLayoutInspectorSection();
          if (datum != null) {
            data.add(new Named<>(datum.getKey(), new StatoObject(datum.getValue())));
          }
        }

        switch (annotation.resType()) {
          case COLOR:
            props.put(f.getName(), fromColor((Integer) f.get(node)));
            break;
          case DRAWABLE:
            props.put(f.getName(), fromDrawable((Drawable) f.get(node)));
            break;
          default:
            if (f.get(node) != null
                && PropWithDescription.class.isAssignableFrom(f.get(node).getClass())) {
              final Object description =
                  ((PropWithDescription) f.get(node)).getStatoLayoutInspectorPropDescription();
              // Treat the description as immutable for now, because it's a "translation" of the
              // actual prop,
              // mutating them is not going to change the original prop.
              if (description instanceof Map<?, ?>) {
                final Map<?, ?> descriptionMap = (Map<?, ?>) description;
                for (Map.Entry<?, ?> entry : descriptionMap.entrySet()) {
                  props.put(entry.getKey().toString(), InspectorValue.immutable(entry.getValue()));
                }
              } else {
                props.put(f.getName(), InspectorValue.immutable(description));
              }
            } else {
              if (isTypeMutable(f.getType())) {
                props.put(f.getName(), InspectorValue.mutable(f.get(node)));
              } else {
                props.put(f.getName(), InspectorValue.immutable(f.get(node)));
              }
            }
            break;
        }
        hasProps = true;
      }
    }

    if (hasProps) {
      data.add(new Named<>("Props", props.build()));
    }

    return data;
  }

  @Nullable
  static StatoObject getStateData(Object node, StateContainer stateContainer) throws Exception {
    if (stateContainer == null) {
      return null;
    }

    final StatoObject.Builder state = new StatoObject.Builder();

    boolean hasState = false;
    for (Field f : stateContainer.getClass().getDeclaredFields()) {
      f.setAccessible(true);

      final State annotation = f.getAnnotation(State.class);
      if (annotation != null) {
        if (DataUtils.isTypeMutable(f.getType())) {
          state.put(f.getName(), InspectorValue.mutable(f.get(stateContainer)));
        } else {
          state.put(f.getName(), InspectorValue.immutable(f.get(stateContainer)));
        }
        hasState = true;
      }
    }

    return hasState ? state.build() : null;
  }

  static boolean isTypeMutable(Class<?> type) {
    if (type == int.class || type == Integer.class) {
      return true;
    } else if (type == long.class || type == Long.class) {
      return true;
    } else if (type == float.class || type == Float.class) {
      return true;
    } else if (type == double.class || type == Double.class) {
      return true;
    } else if (type == boolean.class || type == Boolean.class) {
      return true;
    } else if (type.isAssignableFrom(String.class)) {
      return true;
    }
    return false;
  }

  static com.facebook.stato.plugins.inspector.InspectorValue fromDrawable(Drawable d) {
    if (d instanceof ColorDrawable) {
      return com.facebook.stato.plugins.inspector.InspectorValue.mutable(
          Color, ((ColorDrawable) d).getColor());
    }
    return com.facebook.stato.plugins.inspector.InspectorValue.mutable(Color, 0);
  }

  static com.facebook.stato.plugins.inspector.InspectorValue fromColor(int color) {
    return com.facebook.stato.plugins.inspector.InspectorValue.mutable(Color, color);
  }
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/plugins/litho/DataUtils.kt
package org.stato.plugins.litho



import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import org.stato.core.StatoObject
import org.stato.plugins.inspector.InspectorValue
import org.stato.plugins.inspector.Named
import com.facebook.litho.StateContainer
import com.facebook.litho.annotations.Prop
import com.facebook.litho.annotations.ResType.*
import com.facebook.litho.annotations.State
import org.stato.plugins.inspector.InspectorValue.Type

object DataUtils {

  @Throws(Exception::class)
  internal fun getPropData(node: Any): List<Named<StatoObject>> {
    val props = StatoObject.Builder()
    val data = mutableListOf<Named<StatoObject>>()

    var hasProps = false

    for (f in node.javaClass.declaredFields) {
      f.isAccessible = true

      val annotation = f.getAnnotation(Prop::class.java)
      if (annotation != null) {
        if (f.get(node) != null && PropWithInspectorSection::class.java.isAssignableFrom(f.get(node).javaClass)) {
          val datum = (f.get(node) as PropWithInspectorSection?)?.statoLayoutInspectorSection
          if (datum != null) {
            data.add(Named(datum.key, StatoObject(datum.value)))
          }
        }

        when (annotation.resType) {
          COLOR -> props.put(f.name, fromColor((f.get(node) as Int).toInt()))
          DRAWABLE -> props.put(f.name, fromDrawable(f.get(node) as Drawable))
          else -> if (f.get(node) != null && PropWithDescription::class.java.isAssignableFrom(f.get(node).javaClass)) {
            val description = (f.get(node) as PropWithDescription).statoLayoutInspectorPropDescription
            // Treat the description as immutable for now, because it's a "translation" of the
            // actual prop,
            // mutating them is not going to change the original prop.
            if (description is Map<*, *>) {
              val descriptionMap = description as Map<*, *>
              for (entry in descriptionMap.entries) {
                props.put(entry.key.toString(), InspectorValue.immutable(entry.value ?: continue))
              }
            } else {
              props.put(f.name, InspectorValue.immutable(description))
            }
          } else {
            if (isTypeMutable(f.type)) {
              props.put(f.name, InspectorValue.mutable(f.get(node)))
            } else {
              props.put(f.name, InspectorValue.immutable(f.get(node)))
            }
          }
        }
        hasProps = true
      }
    }

    if (hasProps) {
      data.add(Named("Props", props.build()))
    }

    return data
  }


  @Throws(Exception::class)
  internal fun getStateData(node: Any, stateContainer: StateContainer?): StatoObject? {
    if (stateContainer == null) {
      return null
    }

    val state = StatoObject.Builder()

    var hasState = false
    for (f in stateContainer.javaClass.declaredFields) {
      f.isAccessible = true

      val annotation = f.getAnnotation(State::class.java)
      if (annotation != null) {
        if (isTypeMutable(f.type)) {
          state.put(f.name, InspectorValue.mutable(f.get(stateContainer)))
        } else {
          state.put(f.name, InspectorValue.immutable(f.get(stateContainer)))
        }
        hasState = true
      }
    }

    return if (hasState) state.build() else null
  }

  private fun isTypeMutable(type: Class<*>): Boolean {
    return when(type) {
      Int::class.javaPrimitiveType, Integer::class.java -> true
      Long::class.javaPrimitiveType, Long::class.java -> true
      Float::class.javaPrimitiveType ,Float::class.java -> true
      Double::class.javaPrimitiveType,Double::class.java -> true
      Boolean::class.javaPrimitiveType,Boolean::class.java -> true
      else -> type.isAssignableFrom(String::class.java)
    }
  }

  internal fun fromDrawable(d: Drawable?): InspectorValue<Int> {
    return if (d is ColorDrawable) {
      InspectorValue.mutable(
        Type.Color, d.color)
    } else InspectorValue.mutable(Type.Color, 0)
  }

  private fun fromColor(color: Int?): InspectorValue<Int> {
    return InspectorValue.mutable(Type.Color, color ?: 0)
  }
}
